virtcontainers/fc.go

```diff
@@ -0,0 +1,580 @@
+		return err
+	}
+
+	return nil
+}
+
+func (fc *firecracker) fcSetVMRootfs(path string) error {
+	span, _ := fc.trace("fcSetVMRootfs")
+	defer span.Finish()
+	fc.Logger().WithField("VM rootfs path:", path).Debug()
+
+	driveID := "rootfs"
+	driveParams := ops.NewPutGuestDriveByIDParams()
+	driveParams.SetDriveID(driveID)
+	isReadOnly := false
+	isRootDevice := true
+	drive := &models.Drive{
+		DriveID:      &driveID,
+		IsReadOnly:   &isReadOnly,
+		IsRootDevice: &isRootDevice,
+		PathOnHost:   &path,
+	}
+	driveParams.SetBody(drive)
+	_, err := fc.client.Operations.PutGuestDriveByID(driveParams)
+	if err != nil {
+		fc.Logger().WithField("fcSetVMRootfs failed:", err).Debug()
+		return err
+	}
+
+	return nil
+}
+
+func (fc *firecracker) fcStartVM() error {
+	fc.Logger().Info("start firecracker virtual machine")
+	span, _ := fc.trace("fcStartVM")
+	defer span.Finish()
+
+	fc.Logger().Info("Starting VM")
+
+	fc.client = fc.newFireClient()
+
+	actionParams := ops.NewCreateSyncActionParams()
+	actionInfo := &models.InstanceActionInfo{
+		ActionType: "InstanceStart",
+	}
+	actionParams.SetInfo(actionInfo)
+	_, err := fc.client.Operations.CreateSyncAction(actionParams)
+	if err != nil {
+		fc.Logger().WithField("start firecracker virtual machine failed:", err).Debug()
+		return err
+	}
+
+	fc.state.set(vmReady)
+
+	return nil
+}
+
+// startSandbox will start the hypervisor for the given sandbox.
+// In the context of firecracker, this will start the hypervisor,
+// for configuration, but not yet start the actual virtual machine
+func (fc *firecracker) startSandbox() error {
+	span, _ := fc.trace("startSandbox")
+	defer span.Finish()
+
+	fc.fcInit(FcTimeout)
+
+	kernelPath, err := fc.config.KernelAssetPath()
+	if err != nil {
+		return err
+	}
+
+	strParams := SerializeParams(fc.config.KernelParams, "=")
+	formattedParams := strings.Join(strParams, " ")
+
+	fc.fcSetBootSource(kernelPath, formattedParams)
+
+	image, err := fc.config.InitrdAssetPath()
+	if err != nil {
+		return err
+	}
+
+	if image == "" {
+		image, err = fc.config.ImageAssetPath()
+		if err != nil {
+			return err
+		}
+		//how to handle this error...
+	}
+
+	fc.fcSetVMRootfs(image)
+
+	for _, d := range fc.pendingDevices {
+		if err = fc.addDevice(d.dev, d.devType); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// waitSandbox will wait for the Sandbox's VM to be up and running.
+func (fc *firecracker) waitSandbox(timeout int) error {
+	span, _ := fc.trace("waitSandbox")
+	defer span.Finish()
+
+	return fc.waitVMM(timeout)
+}
+
+// stopSandbox will stop the Sandbox's VM.
+func (fc *firecracker) stopSandbox() error {
+	span, _ := fc.trace("stopSandbox")
+	defer span.Finish()
+
+	fc.Logger().Info("Stopping Sandbox")
+
+	actionParams := ops.NewCreateSyncActionParams()
+	actionInfo := &models.InstanceActionInfo{
+		ActionType: "InstanceHalt",
+	}
+	actionParams.SetInfo(actionInfo)
+	_, err := fc.client.Operations.CreateSyncAction(actionParams)
```

Comment:
I think you need to check `fc.client` and recreate one before using it, in most of the hypervisor calls. Otherwise it might crash for kata cli use case.

Reply:
Fixed that. Ensure that we always check if the socket exist and open it if needed.